### PR TITLE
Add rollout saving to GRPO training

### PR DESCRIPTION
**Pull request by finbarrtimbers**
Implement the save_traces flag to save all generated RL rollouts to disk during GRPO training. For each generation, saves prompt_tokens, response_tokens, advantage, reward, step, sample_idx, prompt_idx, dataset, finish_reason, ground_truth, and tool_info.

- Add rollouts_save_path arg with default /weka/oe-adapt-default/allennlp/deletable_rollouts/
- Save metadata file with git commit, model name, and timestamp
- Save rollouts to sharded JSONL files (new file every 10k samples)
- Use async saving via ThreadPoolExecutor to not block training
- Add `RolloutMetadata` and `RolloutRecord` dataclasses for type safety
- Add GPU integration test validating rollout file creation and field presence

GPU_TESTS=[01KFGWBPSPES7P8X2BP3HGEYGC](https://beaker.org/orgs/ai2/workspaces/open-instruct-dev/work/01KFGWBPSPES7P8X2BP3HGEYGC?taskId=01KFGWBPSXVTFPET6K7XBAGAV2&jobId=01KFGWBQ1S8825KA8C6WZYQMYD)

### CODE REVIEW

File: open_instruct/rl_utils.py [hunk 1]
```diff
@@ -1,16 +1,164 @@
...
+from dataclasses import asdict, dataclass, field
+from typing import TYPE_CHECKING, Generic, TypeVar
 
 import numpy as np
 import torch
 
-from open_instruct import logger_utils
+from open_instruct import data_types, logger_utils, utils
+
+if TYPE_CHECKING:
```
Comment (hamishivi) [inline open_instruct/rl_utils.py]: do we need `if TYPE_CHECKING`?

File: open_instruct/rl_utils.py [hunk 2]
```diff
@@ -1,16 +1,164 @@
...
+@dataclass
+class RolloutMetadata:
+    run_name: str
+    git_commit: str
+    model_name: str
+    timestamp: str
+
+
+@dataclass
+class RolloutRecord:
```
Comment (hamishivi) [inline open_instruct/rl_utils.py:31]: Do you think it would be reasonable to save the logprobs and/or vllm logprobs? One thing I think it would be interesting to do is being able to 'replay' the training run, which I think would need the logprobs?

**Review by hamishivi — COMMENTED** [review]
Generally seems fine!

File: open_instruct/rl_utils.py (hunk 2 above)
Comment (finbarrtimbers) [reply open_instruct/rl_utils.py:31]: Good idea, done!

File: open_instruct/rl_utils.py (hunk 1 above)
Comment (finbarrtimbers) [reply open_instruct/rl_utils.py]: Removed! I also added a pre-commit hook to prevent these.
